chainerrl/agents/td3.py

```diff
@@ -0,0 +1,374 @@
+
+    def sync_target_network(self):
+        """Synchronize target network with current network."""
+        synchronize_parameters(
+            src=self.policy,
+            dst=self.target_policy,
+            method='soft',
+            tau=self.soft_update_tau,
+        )
+        synchronize_parameters(
+            src=self.q_func1,
+            dst=self.target_q_func1,
+            method='soft',
+            tau=self.soft_update_tau,
+        )
+        synchronize_parameters(
+            src=self.q_func2,
+            dst=self.target_q_func2,
+            method='soft',
+            tau=self.soft_update_tau,
+        )
+
+    def update_q_func(self, batch):
+        """Compute loss for a given Q-function."""
+
+        batch_next_state = batch['next_state']
+        batch_rewards = batch['reward']
+        batch_terminal = batch['is_state_terminal']
+        batch_state = batch['state']
+        batch_actions = batch['action']
+        batch_discount = batch['discount']
+
+        with chainer.no_backprop_mode(), chainer.using_config('train', False):
+            next_actions = self.target_policy_smoothing_func(
+                self.target_policy(batch_next_state).sample().array)
+            next_q1 = self.target_q_func1(batch_next_state, next_actions)
+            next_q2 = self.target_q_func2(batch_next_state, next_actions)
+            next_q = F.minimum(next_q1, next_q2)
+
+            target_q = batch_rewards + batch_discount * \
+                (1.0 - batch_terminal) * F.flatten(next_q)
+
+        predict_q1 = F.flatten(self.q_func1(batch_state, batch_actions))
+        predict_q2 = F.flatten(self.q_func2(batch_state, batch_actions))
+
+        loss1 = F.mean_squared_error(target_q, predict_q1)
+        loss2 = F.mean_squared_error(target_q, predict_q2)
+
+        # Update stats
+        self.q1_record.extend(cuda.to_cpu(predict_q1.array))
+        self.q2_record.extend(cuda.to_cpu(predict_q2.array))
+        self.q_func1_loss_record.append(float(loss1.array))
+        self.q_func2_loss_record.append(float(loss2.array))
+
+        self.q_func1_optimizer.update(lambda: loss1)
+        self.q_func2_optimizer.update(lambda: loss2)
+
+    def update_policy(self, batch):
+        """Compute loss for actor."""
+
+        batch_state = batch['state']
+
+        onpolicy_actions = self.policy(batch_state).sample()
+        q = self.q_func1(batch_state, onpolicy_actions)
+
+        # Since we want to maximize Q, loss is negation of Q
+        loss = - F.mean(q)
+
+        self.policy_optimizer.update(lambda: loss)
+
+    def update(self, experiences, errors_out=None):
+        """Update the model from experiences"""
+
+        batch = batch_experiences(experiences, self.xp, self.phi, self.gamma)
+        self.update_q_func(batch)
+        if self.q_func1_optimizer.t % self.policy_update_delay == 0:
+            self.update_policy(batch)
+            self.sync_target_network()
+
+    def select_onpolicy_action(self, obs):
```

Comment:
Is a separate `select_on_policy_action` necessary? Can this not be placed in `act` and `batch_act`?

Reply:
It is just for readability and avoiding duplicate code.